pkg/query/ast.go

```diff
@@ -147,28 +191,121 @@ func (q TopicSelectorNode) GenerateFilter(db *database.Database) database.Filter
 	}
 }
 
-type TopicNode struct {
-	BaseNode
+//-- TimePredicateNode
+
+func (t TimePredicateNode) GenerateFilter(db *database.Database) database.Filter {
+	var startTime, endTime time.Time
+	var err error
+
+	switch t.Value {
+	case "before":
+		endTime, err = t.Children()[0].(*TimeExpressionNode).Time()
+		// It shouldn't be possible there to be an error here (we should catch
+		// it earlier when parsing, so panic
+		if err != nil {
+			panic(err)
+		}
+
+		startTime = db.Segments[0].HeadTime
+	case "since":
+		startTime, err = t.Children()[0].(*TimeExpressionNode).Time()
+		// It shouldn't be possible there to be an error here (we should catch
+		// it earlier when parsing, so panic
+		if err != nil {
+			panic(err)
+		}
+
+		endTime = time.Now()
+	}
+
+	timeRange := database.TimeRange{Start: startTime, End: endTime}
+
+	return func(data database.Entries) database.Entries {
+		if data == nil {
+			return db.Retrieve(database.Query{Range: &timeRange, RangeSemantics: t.Value})
+		}
+
+		// TODO: Handle non-nil case! Let's factor out some of the Retrieve functionality for
+		//       filtering ranges.
+		return nil
+	}
 }
 
-type TimespanNode struct {
-	BaseNode
+//-- TimeExpressionNode
+
+func (t TimeExpressionNode) Time() (time.Time, error) {
+	lh := t.Children()[0].(*TimeWhenceNode)
+	tm, err := lh.Time()
+	if err != nil {
+		return tm, err
+	}
+
+	switch t.Value {
+	case "-":
+		rh := t.Children()[1].(Numeric)
+		return tm.Add(time.Duration(rh.DerivedValue() * -1)), err
+	case "+":
+		rh := t.Children()[1].(Numeric)
+		return tm.Add(time.Duration(rh.DerivedValue())), err
+	}
+
+	return tm, err
 }
 
-func (t TimespanNode) Duration() time.Duration {
+//-- TimeWhenceNode
+
+func (t TimeWhenceNode) Time() (time.Time, error) {
+	switch {
+	case t.Value == "~now":
+		return time.Now(), nil
+	default:
+		return time.Parse(time.RFC3339, t.Value[2:len(t.Value)-1])
+	}
+}
+
+//-- BinaryOpNode
+
+func (b BinaryOpNode) DerivedValue() int64 {
+	lh, rh := b.Children()[0].(Numeric), b.Children()[1].(Numeric)
+
+	switch b.Value {
+	case "*":
+		return lh.DerivedValue() * rh.DerivedValue()
+	case "-":
+		return lh.DerivedValue() - rh.DerivedValue()
+	case "+":
+		return lh.DerivedValue() + rh.DerivedValue()
+	}
+
+	panic(fmt.Sprintf("Unknown operator '%s'", b.Value))
+}
+
+//-- TimespanNode
+
+func (t TimespanNode) DerivedValue() int64 {
 	switch t.Value {
 	case "@year":
-		return time.Hour * 24 * 365
+		return int64(time.Hour * 24 * 365)
```

Comment:
What does this effectively use as a unit? int64 seconds? ms? ns?

Reply:
`time.Duration` are nanoseconds, and is an `int64` under the hood... I wanted an interface so either had to convert everything to a duration, or convert these back to `in64`. I am not sure I made the right decision